### PR TITLE
docs: creating initial document for a finalizer helper

### LINKED ISSUE #460: Any interest in having finalizer helpers in k8util package
I found this handy utility functions when looking through the service catalog code

```
// GetFinalizers gets the list of finalizers on obj
func GetFinalizers(obj runtime.Object) ([]string, error) {
	accessor, err := meta.Accessor(obj)
	if err != nil {
		return nil, err
	}
	return accessor.GetFinalizers(), nil
}

// AddFinalizer adds value to the list of finalizers on obj
func AddFinalizer(obj runtime.Object, value string) error {
	accessor, err := meta.Accessor(obj)
	if err != nil {
		return err
	}
	finalizers := sets.NewString(accessor.GetFinalizers()...)
	finalizers.Insert(value)
	accessor.SetFinalizers(finalizers.List())
	return nil
}

// RemoveFinalizer removes the given value from the list of finalizers in obj, then returns a new list
// of finalizers after value has been removed.
func RemoveFinalizer(obj runtime.Object, value string) ([]string, error) {
	accessor, err := meta.Accessor(obj)
	if err != nil {
		return nil, err
	}
	finalizers := sets.NewString(accessor.GetFinalizers()...)
	finalizers.Delete(value)
	newFinalizers := finalizers.List()
	accessor.SetFinalizers(newFinalizers)
	return newFinalizers, nil
}
```
I find myself copying them to each operator I work on. Any interest in adding them to k8sutil? I happy to take this on

### PR DESCRIPTION
**Pull request by CSdread**
docs: create initial document for finalizer helper
Fixes #460

### CODE REVIEW

**Comment by CSdread** [comment]
Beginning of the finalizer proposal, would like to get it out there to collaborate on

**Comment by CSdread** [comment]
@hasbro17 does this need to be completed before it can be merged?

**Comment by hasbro17** [comment]
@CSdread While we don't need all the implementation details right now, this proposal should have a little more detail on the proposed workflow and abstractions around Finalizers and MutatingAdmissionWebhook.
You can repurpose the discussion from the https://github.com/operator-framework/operator-sdk/issues/460#issuecomment-421381766 for that.

File: doc/proposals/finalizers.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+(finalizer controller). Finalizers cannot be added to resources on creation and require a seperate call to the API to
+add the finalizer for your finalizer controller to the resource. Best practice is to use a MutatingAdmissionWebhook to
+add the finalizer to the resource, you can also add it with an update or patch.
+
+## Goals:
+* Abstract the process of adding and managing finalizers on resources
+* Ease the building of finalizer controller
+* Support abstracting the creation of the MutatingAdmissionWebhook for adding the finalizer
+
+## Details:
```
Comment (shawn-hurley) [inline doc/proposals/finalizers.md]: I would like to see the high level plan. 

As for supporting `MutatingAdmissionWebhook` I wonder if this should be more generic, something someone could use when creating that using the controller-runtime implementation [here](https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg/webhook/admission). Thoughts?

**Comment by CSdread** [comment]
https://github.com/kubernetes-sigs/controller-runtime/issues/140 this issue from controller-runtime is interesting as well, seems they have answers for the webhooks and the admission controllers which is pretty awesome

**Comment by shawn-hurley** [comment]
@CSdread Yeah I think the link in my previous comment is the actual implementation that will be included in the next release. 

I wonder if just defined helpers like #460 are all we want or if we want to make a struct that is specifically used for a single finalizer. 

```go
type FinalizerHelper struct {
    Name string
}

func (f *FinalizerHelper) AddFinalizer(obj runtime.Object) err {
....
}
...

```
Thoughts?

**Comment by CSdread** [comment]
@shawn-hurley i think a finalizer is something that a sdk.Object either can or cannot do, as in my implementation i look at the subject object as the finalizable object based on a constraint interface

**Comment by shawn-hurley** [comment]
@CSdread As we move to controller runtime `sdk.Object` will be used. Considering this move is currently in progress, I think we should plan for this feature in the proposal. Controller runtime does not wrap the `runtime.Object` and the user is responsible for handling the type. That is why I think we should think about something that can handle any kubernetes type of object. Thoughts?

**Comment by pwittrock** [comment]
Instead of using webhooks for this, CRDs support finalizers in their definition.  I think we just need to add support for adding finalizers to the generated CRD

**Comment by CSdread** [comment]
1) we need finalizers on more than CRDs, in my case i need them on pods
2) we can modify my code above to be runtime.Object, either way it works, it still would have to satisfy the interface

**Comment by shawn-hurley** [comment]
Sorry, I am not following the code that you are referencing is it this? 

https://github.com/operator-framework/operator-sdk/issues/460#issuecomment-421381766

**Comment by CSdread** [comment]
yes sorry about that... That is the code we are using for our finalizers, we create them on pods, to do lifecycle events in cassandra before the pod is deleted and GCd

**Comment by shawn-hurley** [comment]
So here is the thing, how does the object, say for an object that you don't own the definition of work? I think what your doing is adding methods to your go structs, which works for your use case, but what if someone is watching pods and wants to add finalizers? This would not work I believe because pods don't ever conform to this interface. 

I think something that handles all types of objects not just generated ones would be worth the more generic options. 

Am I understanding how you are getting this to work correctly?

**Comment by CSdread** [comment]
I believe pods do conform to this interface, along with every other core k8s object that supports finalizers. Anything with a ObjectMeta from v1 meta is going to have it:

https://github.com/kubernetes/apimachinery/blob/ed135c5b96450fd24e5e981c708114fbbd950697/pkg/apis/meta/v1/meta.go#L163

**Comment by CSdread** [comment]
What this does is say, does this kube resource have ObjectMeta, which satisfies the interface, saying this object supports finalizers (they all do)

**Comment by shawn-hurley** [comment]
Oh wow! I didn't even think to go check ObjectMeta. I'm sorry.

I wouldn't mind using parts of this code, I do think that the Finalizer interface should be public and that it does not need to embed the runtime.Object interface.

Could you add a list of these changes in the proposal?

**Comment by CSdread** [comment]
dont be sorry, the API is massive, and i have been knee deep in it for a bit with my cassandra operator. I will add it to the proposal.

**Comment by shawn-hurley** [comment]
@CSdread wondering if you have some commits locally?

File: doc/proposals/finalizers.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Finalizer Helper proposal
+
+## Background:
+In kubernetes, object reclamation on delete is controlled by finalizers. Giving a resource a finalizer will
```
Comment (jmrodri) [inline doc/proposals/finalizers.md]: * Change kubernetes to Kubernetes.

**Comment by shawn-hurley** [comment]
Closing this PR as there has been no action on it and we are working on solutions in other PRs.

**Pull request closed without merging**